local/validator.go

```diff
@@ -0,0 +1,23 @@
+package local
+
+import (
+	"fmt"
+	"strconv"
+)
+
+func validateMode(i interface{}, k string) (s []string, es []error) {
+	v, ok := i.(string)
+
+	if !ok {
+		es = append(es, fmt.Errorf("expected type of %s to be string", k))
+		return
+	}
+
+	fileMode, err := strconv.ParseInt(v, 8, 64)
+
+	if err != nil || fileMode > 0777 {
+		es = append(es, fmt.Errorf("bad mode for file: %s", v))
```

Comment:
Can save this for a later update if needed, but I think this message isn't likely to be very actionable since it's including a raw error message from `strconv.ParseInt` and those tend to not be suitable for our end-user audience.

Instead, I'd suggest a custom message that more directly describes the problem, like:

```
Invalid file mode: must be a three octal digits
```

---

While on this subject, we could perhaps also be a little stricter in what format we accept here considering that it's likely to be a mistake if this string isn't either exactly three octal digits or four digits where the first one is zero. Part of me thinks that we might be better off not even allowing that leading zero (it's redundant anyway) but putting that aside, a value like `7` would pass validation here but it doesn't seem like that is something a user would ever _intentionally_ do. What do you think? :thinking:

Reply:
Leading zero with a string is what I've seen most people use, esp in the infra management world (eg. Puppet, Chef, Salt, Ansible all use a leading zero). And the Golang example uses leading zero as well: https://golang.org/pkg/os/#Chmod

On the validation side, I've added some specifics to make sure it's a 3 or 4 character string. Technically "7" is a valid number for chmod without the leading zero's, it just treats it as 0007:

```
$ chmod 7 foo.txt
$ ll foo.txt
-------rwx  1 psouter  staff  0 26 Jul 23:16 foo.txt
```

But lets go a bit stricter because that's a weird usecase and probably a footgun and restrict to 3 or 4 length strings.

Reply:
Looks good, thanks!

Indeed, for the length thing I was thinking it better to be more restrictive than what `chmod` might accept in the interests of giving better feedback for a user error, rather than just silently doing something "odd".

Reply:
👍